Implementation of Table Index Engine using sortedcontainers

grantjenks opened the pull request:
This is an implementation of a table index engine using sortedcontainers. It does not yet pass all tests. It's based on the FastRBT engine in bst.py. My thought for moving forward would be to introduce a new engine like SortedArray and then deprecate the old engines based on bintrees.

Using the bintrees design, I replaced bintrees.FastRBTree with sortedcontainers.SortedDict. The flaw here is that SortedDict requires that keys be hashable and the Time object doesn't support hashing though it supports comparisons. These changes add support for hashing based on #7532

There's also a fix for the hash function in astropy.units.Quantity. Previously it xor'd the units but that requirement is not checked in equality. Example with old changes:

```
In [6]: from astropy.units import *

In [8]: Quantity(1.0) == 1.0
Out[8]: True

In [9]: hash(Quantity(1.0)) == hash(1.0)
Out[9]: False
```

Reference: Issue #6539. Previous pull request #7531. Time hash function modified from #7532.

grantjenks commented:
I have pytest passing locally with one additional change: time comparisons, `Time.__lt__`, now returns `NotImplemented` on mismatched types. Previously it would raise `OperandTypeError` on mismatched types. It's needed for the indexing API, which uses `MinValue` and `MaxValue` for range queries. I suspect `bintrees` always compared the values as `MinValue() < Time(...)` and so always used `MinValue.__lt__`. Sorted Containers uses the `bisect` module which instead compares using `Time.__lt__` and so generates `OperandTypeError` without giving `MinValue.__lt__` a chance at the comparison. Python has a mechanism for supporting this scenario: [`NotImplemented`](https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy). By returning `NotImplemented`, Python will attempt the operation again with the operands reflected. I only made the change to `__lt__` to observe the impact with minimal changes. I'm not sure how much code relies on `OperandTypeError`.

As I've thought more about hashing time objects, I'm concerned that what I have is not quite right. In fact, you may have a type for which a hash method is impossible. Since time objects with different scales may be equal, the hash would need to convert the scale to something universal. Is there such a thing?

grantjenks commented:
Last big update. The change here is to use a SortedList rather than SortedDict. The benefit of SortedList is that it does not require hashing. With these changes, Time objects *do not need* to support hashing. Although comparison with Time objects *does need* to return NotImplemented for unsupported operand types.

taldcroft commented on astropy/table/tests/test_index.py:
```diff
@@ -20,7 +21,7 @@
 
 if HAS_BINTREES:
-    available_engines = [BST, FastBST, FastRBT, SortedArray]
+    available_engines = [BST, FastBST, FastRBT, SortedArray, SCEngine]
```
This block should be more like:
```
available_engines = [BST, SortedArray]
if HAS_BINTREES:
    available_engines.extend([FastBST, FastRBT])
if HAS_SOCO:
    available_engines.append(SCEngine)
```

taldcroft commented on astropy/time/core.py:
```diff
@@ -789,6 +789,9 @@ def __setitem__(self, item, value):
         self._time.jd1[item] = value._time.jd1
         self._time.jd2[item] = value._time.jd2
 
+    def __hash__(self):
```
Remove if not needed.

mhvk replied:
This hash is also incomplete (should include scale and location), so either adjust or remove.

taldcroft commented on astropy/table/soco.py:
```diff
@@ -0,0 +1,173 @@
+"""
+
+from collections import OrderedDict
+from sortedcontainers import SortedList
```
This module needs to still be importable even if `sortedcontainers` is not installed.  Then if user tries to create an `SCEngine` instance it should raise ImportError (probably?) at that point with a message that `sortedcontainers` needs to be installed.  This is slightly different than the BST version, where there is always the built-in backup if no `bintrees` is available.

Standard idiom is to define a `HAS_SOCO` module global in a try import / except block.  You can then import that in testing to decide whether or not to run index tests with `SCEngine`.

grantjenks replied:
This seems strange to me. There can be no SCEngine without `sortedcontainers`. To cause the import error when the `SCEngine` is created, the import would have to be moved to its `__init__` method.

The BST version is also a bit odd to me because if you try to import `FastBST` without `bintrees` installed then it lies and gives `BST` instead. What's the use-case here: to allow code that uses `FastBST` to work without `bintrees`? How would those users become aware of the degraded performance? Maybe there should be a warning emitted?

Another idea, to implement similar functionality here, would be to move the `from sortedcontainers import SortedList` line to the bottom in a try/except such that if we catch an `ImportError` then we set `SCEngine = BST` or whatnot. Again, probably should emit a warning.

I'm happy to see it implemented whatever way. Just want to understand better.

MSeifert04 replied:
While I agree that it would be odd to have different APIs depending on the installed packages I also think that *silently* falling back to slower (and possibly differently behaving) implementations isn't great either.

So I'd say the module should always be importable and the SCEngine should always be present in the API we should actually error out or warn if it's used without having sortedcontainers installed.

grantjenks replied:
Ok, I defined HAS_SOCO and made the module always importable. If Sorted Containers is not installed then SCEngine is present but cannot be instantiated.

taldcroft commented on astropy/table/tests/test_index.py:
```diff
@@ -6,6 +6,7 @@
 from .test_table import SetupData
 from ..bst import BST, FastRBT, FastBST
 from ..sorted_array import SortedArray
+from ..soco import SCEngine
```
```
from ..soco import SCEngine, HAS_SOCO
```

taldcroft commented:
@grantjenks - from a quick look this seems great.  I can't think of any immediate problem with changing the time comparison exception as you have done.  Will think about it.

Is the change to the Quantity hash still required for this Table indexing update?  The change you have made is likely not OK since you want values with different units to have a different hash:
```
In [10]: hash(1 * u.m) == hash(1 * u.km)
Out[10]: False
```
Of course you get into the same conundrum with Time, where Quantities that look different do compare equal:
```
In [12]: 1000 * u.m == 1 * u.km
Out[12]: True

In [13]: hash(1000 * u.m) == (1 * u.km)
Out[13]: False
```
From the Quantity perspective `1000 * u.m` is a different thing from `1 * u.km`, so it makes sense to have the hash be different.  I think the hash rule that objects which compare equal should have the same hash is a bit too simplistic.

taldcroft commented:
@mhvk - please see the [comment above](https://github.com/astropy/astropy/pull/7574#issuecomment-398475373) about changing all the Time comparisons to raise `NotImplemented` instead of `OperandTypeError`.  I think it makes sense, but since you implemented this, you should weigh in.

taldcroft commented:
This will also need two entries in the `CHANGES.rst` file:
- New feature in `astropy.table` for the new engine
- API change in `astropy.time` for the Time comparison exception change

mhvk commented on astropy/units/quantity.py:
```diff
@@ -1149,7 +1149,7 @@ def __pos__(self):
     # other overrides of special functions
     def __hash__(self):
-        return hash(self.value) ^ hash(self.unit)
+        return hash(self.value)
```
As noted by @taldcroft, definitely not OK to change...

mhvk commented on astropy/time/core.py:
```diff
@@ -1653,7 +1656,10 @@ def _time_difference(self, other, op=None):
 
     def __lt__(self, other):
-        return self._time_difference(other, '<') < 0.
+        try:
```
It is indeed better to return `NotImplemented`, though it should be done in `_time_difference` - I'll make a quick PR that this can be rebased on.

mhvk commented:
@taldcrof - I'll do the `Time` change separately (where I can be a bit more logical and do the same for `add` and `sub`).

mhvk commented:
@grantjenks - see #7584. Possibly easiest if we get that in quickly and then you can rebase on master to have it included.

For the change to `Quantity.__hash__`, if you cannot do without it, could you explain what goes wrong?

grantjenks commented:
@mhvk I don't need the `Quantity.__hash__` change. Sorry, I should've removed that.

I will wait on #7584 and then rebase.

grantjenks commented:
I've removed the `__hash__` changes and included commits from @mhvk (just to see it pass CI).

I think all that remains is to:

1. Rebase on top of @mhvk changes once they're in `master`.
2. Settle on what the imports for `test_index.py` should look like. (Please review.)
3. Decide if `astropy.table.soc` must always be importable. See comment with @taldcroft above.
4. Add whatever necessary to `CHANGES.rst`

taldcroft commented on astropy/table/index.py:
```diff
@@ -243,7 +243,7 @@ def remove_row(self, row, reorder=True):
         '''
         # for removal, form a key consisting of column values in this row
-        if not self.data.remove(tuple([col[row] for col in self.columns]), row):
+        if not self.data.remove(tuple(col[row] for col in self.columns), row):
```
Agreed that it is generally better to leave unrelated code alone, but in this case it is already done and it is indeed an improvement.

MSeifert04 commented:
Just a Note even if we do that in another  pr: we have to add soco to the optional dependencies list and test matrix.

grantjenks commented:
I updated the pull request and added changes to a couple places where I found references to bintrees.

grantjenks commented:
I added a reference to SCEngine in a few more places where I also found references to FastBST and FastRBT.

I also found some previous benchmarking and updated it to include SCEngine: https://github.com/grantjenks/astropy-notebooks/blob/master/table/indexing-profiling.ipynb The benchmarking found a couple opportunities for speedups so now initializing and removing rows is faster. The benchmarks also show sortedcontainers as on-par or faster than bintrees.

bsipocz commented on astropy/table/soco.py:
```diff
@@ -0,0 +1,170 @@
+class SCEngine:
+    '''
+    Fast tree-based implementation for indexing, using the
+    `sortedcontainers` package.
```
Could you use double backticks otherwise sphinx fails when tries to create a link for this.

taldcroft commented:
@grantjenks - I pushed a little commit to fix the docstring error and am merging now.  Thanks so much for this great contribution!  You might note that I've added a `whatsnew-needed` tag, which means I think this is important enough to get a callout in the next release highlights.

grantjenks commented:
Been on vacation but just wanted to circle back and say: thanks to you all too! I'm always glad to see Sorted Containers used in more places and happy to support the Astropy project.

Keep up the fascinating work!
